app/services/merge_patient_service.rb

```diff
@@ -1,64 +1,61 @@
+    patient_attributes =
+      payload
+        .except(:address, :phone_numbers, :business_identifiers)
+        .yield_self { |params| set_metadata(params) }
+        .yield_self { |params| set_address_id(params, merged_address) }
+        .yield_self { |params| set_assigned_facility(params) }
 
-    patient_attributes["address_id"] = merged_address.id if merged_address.present?
-    merged_patient = Patient.merge(attributes_with_metadata(patient_attributes))
+    merged_patient = Patient.merge(patient_attributes)
     merged_patient.address = merged_address
-
     merged_phone_numbers = merge_phone_numbers(payload[:phone_numbers], merged_patient)
-    merged_business_identifiers = merge_business_identifiers(payload[:business_identifiers], merged_patient)
-
-    if (merged_address.present? && merged_address.merged?) || merged_phone_numbers.any?(&:merged?) || merged_business_identifiers.any?(&:merged?)
-      merged_patient.touch
-
-      #
-      # This is a rare scenario that might be possible in the future.
-      # If the client allows the user to update the patient's address or phone_number,
-      # there might be a case where the address or phone_number for a patient is updated for a discarded patient.
-      #
-      # These updates should not ideally be made at all because they will be invisible to the user.
-      #
-      # We can fix this issue, but it requires re-working the merge function, so we'll currently just
-      # track the incidence rate, so we can plan for a fix if necessary.
-      log_update_discarded_patient(merged_patient)
-    end
+    merged_business_ids = merge_business_identifiers(payload[:business_identifiers], merged_patient)
 
-    if merged_patient.deleted_at.present? && existing_patient_attributes&.dig("deleted_at").nil?
-      # Patient has been soft-deleted by the client, server should soft-delete the patient and their associated data
-      # patient_attributes[:metadata][:registration_user_id] contains the current user's id
-      merged_patient.update(deleted_by_user_id: patient_attributes[:metadata][:registration_user_id])
-      merged_patient.discard_data
-    end
+    touch_patient(merged_patient, merged_address, merged_phone_numbers, merged_business_ids)
+    discard_patient_data(merged_patient)
 
     merged_patient
   end
 
   private
 
-  def attributes_with_metadata(patient_attributes)
-    with_request_metadata = patient_attributes.merge(patient_attributes[:metadata]).except(:metadata)
-    metadata_keys = patient_attributes[:metadata].keys
-
-    case Patient.compute_merge_status(with_request_metadata)
-    when :new
-      with_request_metadata
-    when Set[:updated, :old]
-      existing_metadata = Patient.find(patient_attributes[:id]).slice(*metadata_keys)
-      patient_attributes.merge(existing_metadata).except(:metadata)
-    else
-      patient_attributes.except(:metadata)
+  attr_reader :existing_patient, :payload, :request_metadata
+
+  def set_metadata(patient_params)
```

Comment:
This is really the method that has had the most impact with this refactor.

We keep the behavior where new patients get their metadata updated through either the headers or through the params and the patients being updated don't. We can change this behavior later easily when we want to update `registration_facility` or `registration_user`.

However, here now, this doesn't do any weird metadata pulling/de-structuring, it just refers to the `new_patient_metadata` or `existing_patient_metadata` functions respectively as needed.